staging/src/k8s.io/kubectl/pkg/cmd/wait/wait.go

```diff
@@ -152,6 +153,7 @@ func (flags *WaitFlags) AddFlags(cmd *cobra.Command) {
 
 	cmd.Flags().DurationVar(&flags.Timeout, "timeout", flags.Timeout, "The length of time to wait before giving up.  Zero means check once and don't wait, negative means wait for a week.")
 	cmd.Flags().StringVar(&flags.ForCondition, "for", flags.ForCondition, "The condition to wait on: [delete|condition=condition-name[=condition-value]|jsonpath='{JSONPath expression}'=[JSONPath value]]. The default condition-value is true.  Condition values are compared after Unicode simple case folding, which is a more general form of case-insensitivity.")
+	cmd.Flags().BoolVar(&flags.WaitForCreation, "wait-for-creation", flags.WaitForCreation, "Default behavior of wait command errors out when the resource does not exist. If this flag is specified, command also waits the creation of the resource, in case it doesn't exist.")
```

Comment:
Was it considered to just add `create` as another condition on the `--for` flag?

To me, this seems logical, since we already have a `delete` condition, which it would be the opposite of.

So something like...
```
kubectl wait --for=create pod/nginx
```

If you really want to wait for creation and then wait for the pod to become ready, you could accomplish that with two waits:
```
kubectl wait --for=create pod/nginx
kubectl wait --for=condition=Ready pod/nginx
```

I'm not opposed to having it as a separate flag, just wondering if there is some advantage to treating it as a regular condition instead of a pre-condition that has to be used in conjunction with another condition.

Reply:
`--for=create` would be a better idea because it aligns with the other for conditions (e.g. `--for=delete`) as you said and code implementation also would be easier for me. 

However, after thinking a bit on this, `--for=delete` is apparently mutually exclusive with the other conditions and it makes sense to have it separately. Because people would not want to wait for deletion of a resource simultaneously with the other conditions. On the other hand, it seems that creation is embedded to other conditions and people want to also wait creation while waiting for their conditions in the same command.

If there is any scenario that `--for=creation` would be useful without needing the `--for=condition`, I think it makes sense to have it separately. WDYT?.

Reply:
> If there is any scenario that --for=creation would be useful without needing the --for=condition, I think it makes sense to have it separately. WDYT?.

I don't know if there is any specific scenario, but I was thinking it would be more consistent for it to be a `--for` condition, and that would allow it to be used independently if needed.

But I don't want to over-complicate things either.

I could go either way.  I think the important thing is to get this into kubectl so people don't have to write workarounds.  I'd be interested in hearing other's takes on it, but if nobody else wants to opine, then I will say whatever way you decide is fine.

---

Another thought...

I can't think of a time when I would run `kubectl wait --for=condition=Ready pod/foo` that I would want it to fail immediately if the pod doesn't exist (yet).

Do you foresee this becoming the default behavior of `kubectl wait` in a future version?

If so, when people use `kubectl wait --for=condition=XXXX` without specifying `--wait-for-creation=false`, we could print a warning telling them that in a future version, kubectl wait will by default wait for creation before waiting for a condition.  Then after the deprecation period, we could switch the default behavior to always wait for creation by default (unless you are waiting for deletion of course).

WDYT?

Reply:
Thanks for dropping your thoughts. I've added this topic in sig-cli biweekly meeting on March of 6th 2024 to discuss.

Reply:
@eddiezane what is your opinion about this after our last discussion in cli bi-weekly?.

Reply:
I talked to a few different folks and I think that if we go with the flag route that we should default `--wait-for-creation=true`.

Most folks (especially in a GitOps/CI env) are either running 2 wait's or doing loop/retry logic to get around the current behavior of failing if the resource isn't created.

> In my experience it has always been a multi-step process. Some loop to wait until it exists, then the `kubectl wait` to wait for whatever condition is desired, so this sounds awesome

> kubectl wait implicitly waiting for creation by default would make my life so much easier

Examples:
- https://github.com/tektoncd/dashboard/blob/10b1c8434ffc9719f21fbe744030ff741e691c4b/scripts/prepare-kind-cluster#L67-L87
- https://github.com/minio/operator/blob/c2988daba81534f14c116fc74cf2d49c4f920fcb/tests/common.sh#L58-L65
- https://github.com/weaveworks/weave-gitops/blob/1616603c008eb0e392157fa135f662570a7b11db/tools/kind-with-ingress.sh#L32-L43
- https://github.com/ongres/stackgres/blob/663c118978e6203ac82cea7685e241b9f4566827/stackgres-k8s/e2e/utils/dbops#L175-L208
- https://github.com/labring/sealos/blob/3859de964955dda9a350ddbc5cda7f9f48c2cc6a/deploy/objectstorage/scripts/minio.sh#L34-L36
- https://github.com/apigee/devrel/blob/a94de3b3b65c6780c21af7e12901eb6a205e6926/tools/hybrid-quickstart/steps.sh#L717
- https://github.com/defenseunicorns/zarf/blob/main/src/pkg/utils/wait.go
- https://github.com/saferwall/saferwall/blob/f1065a0c53fa88bbfb768ec47d3f001a7a544749/build/mk/kind.mk#L18

I haven't been able to surface anything that demands the current behavior so I think that defaulting to true and allowing whatever special cases come up to explicitly set `--wait-for-creation=false` sounds good.

Thoughts?

Reply:
Thank you for your comment. All the links you gave ^ try to solve this by adding sleeps and I agree that flag route will simplify these scripts by taking away the prerequisite of waiting for creation additionally. 

As changing the default behavior to `--wait-for-creation=true` covers all the cases that the current default behavior has, I think we are fine in terms of backwards compatibility and we can change the default as well. 

@brianpursley @mpuckett159 WDYT?